### DISCUSSION TITLE
Receive BLE sensor Characteristic information

### DISCUSSION THREAD

**Discussion by j-franco7** [hbldh/bleak#722] [Q&A]
- Base Device: Raspberry Pi 4 Model B

- bleak version: 0.13.0

- Python version:  3.9.2

- Operating System: Raspbian GNU/Linux 11

- BlueZ version: 5.55

I am doing a university work where it is necessary through a Raspberry Pi to access information that is reported by a blueetooth sensor that has an IMU unit (with accelerometer, gyroscope and magnetometer).

Before starting to build code in python on the Raspberry Pi, I installed an Android application called nRF Connect that allows you to access information from blueetooth devices.

After successfully connecting to the Bluetooth device, I was able to access a number of bluetooth characteristics, as can be seen from the figure below:

![1](https://user-images.githubusercontent.com/90055042/148526084-74a79c49-8ec6-4312-a533-9eebbe52464c.jpg)

The instructions I have on this sensor is that in order to access the values indicated by the gyroscope, accelerometer and magnetometer, we need to send the 0x01 byte array via UUID 0xFF35. So in the nRF Connect application I wrote 0x01 to the characteristic 0xFF35:

![2](https://user-images.githubusercontent.com/90055042/148527019-8cc46f01-25a2-4e36-8cbc-694940cf6332.jpg)

After sending this byte 0x01 to UUID 0xFF35, the characteristic with UUID 0xFF38 immediately starts sending the required information. This value is constantly changing as it refers to the three sensors that are installed in the IMU unit. As can be seen by the red box in the figure below:

![3](https://user-images.githubusercontent.com/90055042/148527523-cc18a425-6509-4bb1-b09a-20d9d0957ca9.jpg)

Next, I tried replicating the same procedures to access the values of characteristic 0xFF38, using python code that uses functions from the Bleak project. That is, I wrote byte 0x01 at characteristic 0xFF35 and tried to read from characteristic 0xFF38:

The python code that I composed is as follows:

```python
import asyncio
import bleak

DEVICE_MAC = '20:C3:8F:D0:CF:10' # MAC of BLE Device
CONFIG_UUID = '0000ff35-0000-1000-8000-00805f9b34fb' # Change this to your characteristic UUID
CONFIG_UUID2 = '0000ff38-0000-1000-8000-00805f9b34fb'

async def run():
    # Connect to client and wait for connection to complete
    client = bleak.BleakClient(DEVICE_MAC, timeout=10)
    await client.connect()

    # Characteristic 0xFF35 before it is set
    print(f'Before write 0x01 in Characteristic 0xFF35 : {await client.read_gatt_char(CONFIG_UUID)}')

    # Send 0x01 byte to 0xFF35
    data = b'\x01' 
    await client.write_gatt_char(CONFIG_UUID, data, True)

    # Characteristic 0xFF35 after I wrote byte 0x01:
    print(f'After write 0x01 in Characteristic 0xFF35 : {await client.read_gatt_char(CONFIG_UUID)}')
   
    await asyncio.sleep(1.0)
   
   # Read values reported by 0xFF38:
    print(f'In characteristic 0xFF38: {await client.read_gatt_char(CONFIG_UUID2)}')

if __name__ == "__main__":
    # Run the program
    loop = asyncio.get_event_loop()
    loop.run_until_complete(run())  

```
When I run this program on the Raspeberry Pi, I get the following result:

![4](https://user-images.githubusercontent.com/90055042/148533182-695ecdbd-f180-4717-bde3-68889cd75fe6.JPG)

Considering the last image, I can't access the values read by characteristic 0xFF38, as I could through the nRF Connect application, since I get the error indicated in the image.

On the other hand, I tried running the program "service_explorer.py" (https://github.com/hbldh/bleak/blob/develop/examples/service_explorer.py) and it shows all the services and characteristics of the BLE sensor described.

However, the characteristic 0xff38, shows the same error, as shown in the message below:

![5](https://user-images.githubusercontent.com/90055042/148535039-9ca1c2d3-79e4-4683-bae6-b52003a5fa73.jpg)

-With this in mind, what could be the main cause of not having access to the information sent through characteristic 0xFF38?

Thank you for your attention.

**Comment by dlech** [comment]
This error is coming from the Bluetooth chip itself. I would try searching the web for that error message and your Bluetooth chip to see what you can find.

**Reply by j-franco7** [reply]
Good Morning,

In fact what is strange about this situation is that through the nRF Connect app I can check the values of the characteristic 0xFF38 to vary, on the other hand on the Raspberry Pi side, I am faced with this error.
Thank you for your attention.
